### PR TITLE
Fix TextureCube using promises as data

### PR DESCRIPTION
**Pull request by georgios-uber**

### CODE REVIEW

File: modules/core/src/scenegraph/gltf-instantiator/gltf-material.js [hunk 1]
```diff
@@ -75,25 +75,16 @@ class GLTFEnv {
...
-        pixels: {
-          [this.gl.TEXTURE_CUBE_MAP_POSITIVE_X]: faces[0],
-          [this.gl.TEXTURE_CUBE_MAP_POSITIVE_Y]: faces[1],
-          [this.gl.TEXTURE_CUBE_MAP_POSITIVE_Z]: faces[2],
-          [this.gl.TEXTURE_CUBE_MAP_NEGATIVE_X]: faces[3],
-          [this.gl.TEXTURE_CUBE_MAP_NEGATIVE_Y]: faces[4],
-          [this.gl.TEXTURE_CUBE_MAP_NEGATIVE_Z]: faces[5]
-        }
-      });
+    return new TextureCube(this.gl, {
```
Comment (ibgreen) [inline modules/core/src/scenegraph/gltf-instantiator/gltf-material.js]: Nit: Can we use a more descriptive name than `func`? `getTextureForFace`?

File: modules/core/src/scenegraph/gltf-instantiator/gltf-material.js [hunk 2]
```diff
@@ -75,25 +75,16 @@ class GLTFEnv {
...
+      id,
+      pixels: {
+        [this.gl.TEXTURE_CUBE_MAP_POSITIVE_X]: func('right'),
```
Comment (ibgreen) [inline modules/core/src/scenegraph/gltf-instantiator/gltf-material.js]: Are these names part of the standard in any way? (right, left, ...)? 

Or just an artifact of the examples we happen to use?

If not, can we send in something more generic? E.g. the WebGL constant?

File: modules/core/src/webgl/classes/texture-cube.js [hunk 3]
```diff
@@ -13,27 +13,31 @@ const FACES = [
...
+    // let {width = 1, height = 1, type = GL.UNSIGNED_BYTE, dataFormat} = props;
 
     // Deduce width and height based on one of the faces
-    ({type, dataFormat} = this._deduceParameters({format, type, dataFormat}));
-    ({width, height} = this._deduceImageSize({
-      data: props[GL.TEXTURE_CUBE_MAP_POSITIVE_X],
-      width,
-      height
-    }));
+    // ({type, dataFormat} = this._deduceParameters({format, type, dataFormat}));
```
Comment (ibgreen) [inline modules/core/src/webgl/classes/texture-cube.js]: Can deduction be restored?

File: modules/core/src/webgl/classes/texture-cube.js [hunk 4]
```diff
@@ -46,16 +50,18 @@ export default class TextureCube extends Texture {
...
-      this.generateMipmap(props);
-    }
-
     // Store props for accessors
     this.opts = props;
+
+    this.setCubeMapImageData(props).then(() => {
+      this.loaded = true;
+
+      // TODO - should genMipmap() be called on the cubemap or on the faces?
```
Comment (ibgreen) [inline modules/core/src/webgl/classes/texture-cube.js]: Great if we can research this before we land?

Comment (tsherif) [reply modules/core/src/webgl/classes/texture-cube.js]: `genMipamap` should be called once on the cubemap. It's a per-texture operation.

File: modules/core/src/scenegraph/gltf-instantiator/gltf-material.js (hunk 2 above)
Comment (tsherif) [reply modules/core/src/scenegraph/gltf-instantiator/gltf-material.js]: +1

Also, nit: I generally prefer ordering sequential constants in the order of the values:
screen shot 2019-02-25 at 11 24 19 am
Makes it easier to switch to an array if that ever happens.

File: modules/core/src/webgl/classes/texture-cube.js [hunk 5]
```diff
@@ -46,16 +50,18 @@ export default class TextureCube extends Texture {
...
+      // TODO - without generateMipmap() cube textures do not work at all!!! Why?
```
Comment (tsherif) [inline modules/core/src/webgl/classes/texture-cube.js]: Are you getting errors here? And does Luma use immutable textures? If you allocate storage beforehand and put aside space for mipmaps, then the texture would be incomplete before mipmaps are generated.

Comment (ibgreen) [reply modules/core/src/webgl/classes/texture-cube.js]: > does Luma use immutable textures?

We haven't started using immutable texture, but it is definitely of interest, and the API should support it

> the texture would be incomplete before mipmaps are generated.

Good point. We do want an API that enables (and ideally makes it convenient) to work with such use cases.

File: modules/core/src/webgl/classes/texture-cube.js (hunk 3 above)
Comment (georgios-uber) [reply modules/core/src/webgl/classes/texture-cube.js]: Well it will have to be another promise there...

File: modules/core/src/webgl/classes/texture-cube.js (hunk 4 above)
Comment (georgios-uber) [reply modules/core/src/webgl/classes/texture-cube.js]: @tsherif can you explain?

Comment (georgios-uber) [reply modules/core/src/webgl/classes/texture-cube.js]: Why does this code work anyway?

File: modules/core/src/webgl/classes/texture-cube.js (hunk 5 above)
Comment (georgios-uber) [reply modules/core/src/webgl/classes/texture-cube.js]: But still why does it not work if we don't call **generateMipmap**?

Comment (tsherif) [reply modules/core/src/webgl/classes/texture-cube.js]: @georgios-uber It's because the default value of MIN_FILTER is NEAREST_MIPMAP_LINEAR (https://www.khronos.org/registry/OpenGL-Refpages/es3.0/html/glTexParameter.xhtml).

If you try to draw with that min filter without generating the mip maps, the texture will be considered incomplete. 

The clue is following error being thrown in the console: `RENDER WARNING: texture bound to texture unit 0 is not renderable. It maybe non-power-of-2 and have incompatible texture filtering.`

Comment (tsherif) [reply modules/core/src/webgl/classes/texture-cube.js]: I'd recommend not allowing `generateMipmap` to be passed in as an argument, and instead deducing it from the min filter ([e.g.](https://github.com/tsherif/picogl.js/blob/master/src/texture.js#L114)).

Comment (georgios-uber) [reply modules/core/src/webgl/classes/texture-cube.js]: I tried changing the filter to LINEAR and did not fix the issue. Let me try again.

Comment (tsherif) [reply modules/core/src/webgl/classes/texture-cube.js]: BTW, are you using SpectorJS? https://spector.babylonjs.com/

It's a godsend for debugging stuff like this.

**Comment by georgios-uber** [comment]
Ping ping.
